### ISSUE TITLE
Local type variable is inferred as union, "forgetting" constrained TypeVar

### ISSUE THREAD

**Issue by nh2** [microsoft/pyright#5448]
On `pyright 1.1.309`:

```python
from typing import TypeVar
from pathlib import Path

T = TypeVar("T", Path, float)

def f(x: T, y: T):
    # * pyright outputs: `Type of "x" is "T@f"`
    reveal_type(x)

    a: T = x / y

    # * pyright outputs: ` Type of "a" is "Path* | float*"`, should it not also be `T@[f]`?
    reveal_type(a)

    f(x, a)  # why does this typecheck, given that apparently pyright has forgotten that `a` is of type `T@[f]`?`
```

* Why does pyright conclude that `Path* | float*`?
* I also find it slightly problematic that it infers that even though I explicitly write `a: T`, and this conflict is no error.
* How does `f(x, a)` typecheck?

In my application (from which the above is a minimal extract) I am apparently not so lucky as with `f(x, a)` here, and I get a type error later when using `a`.

## Workaround

```python
def f(x: T, y: T):

    def compute_a() -> T:
        return x / y

    a: T = compute_a()

    reveal_type(a)  # As expected: `Type of "a" is "T@f"`
```

This seems inconvenient, to have to wrap all my local-variable computations into helper functions to get the type inference correct.

Maybe I'm missing something, but it feels like a bug.

Thanks!

**Comment by nh2** [comment]
I should say, if I use

```diff
-def compute_a() -> T:
+def compute_a():
```

then the workaround does not work.

**Comment by erictraut** [comment]
Pyright is working correctly here. Please refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=conditional-types-and-type-variables) for details.
